### PR TITLE
ensure proper hierarchy when creating btrfs subvolumes (bsc#1078732)

### PR DESCRIPTION
**Pull request by wfeldt**
When creating btrfs subvolumes the proper hierarchy must be preserved. It's not possible to create `/foo/bar` when a subvolume `/foo/bar/xxx` already exists.

If the filesystem hasn't been created yet we can just rearrange all subvolumes to fit the new one in. For an existing filesystem there's nothing we can do except notify the user.

### CODE REVIEW

File: src/lib/y2storage/filesystems/btrfs.rb [hunk 1]
```diff
@@ -119,7 +149,7 @@ def delete_btrfs_subvolume(devicegraph, path)
         devicegraph.remove_btrfs_subvolume(subvolume)
       end
 
-      # Creates a new btrfs subvolume for the filesystem
+      # Create a new btrfs subvolume for the filesystem
```
Comment (ancorgs) [inline src/lib/y2storage/filesystems/btrfs.rb]: Any reason to change this? In fact, the whole class (except the methods you added and this one you changed) are documented in third person (i.e. with the "s"), following the Ruby convention.

It's not a big deal if some people prefer to not follow the convention when they document a new method. But I would avoid a war in which each side "fixes" the wording used by the other side on every modification of a documented method. :wink:

File: src/lib/y2storage/filesystems/btrfs.rb [hunk 2]
```diff
@@ -128,12 +158,23 @@ def delete_btrfs_subvolume(devicegraph, path)
...
+      #
+      # @return [BtrfsSubvolume, nil] new subvolume
+      #
       def create_btrfs_subvolume(path, nocow)
-        subvolume = default_btrfs_subvolume.create_btrfs_subvolume(path)
-        subvolume.nocow = nocow
-        subvolume_mount_path = btrfs_subvolume_mount_point(path)
-        subvolume.create_mount_point(subvolume_mount_path) unless subvolume_mount_path.nil?
-        subvolume
+        subvolume = find_btrfs_subvolume_by_path(path)
```
Comment (ancorgs) [inline src/lib/y2storage/filesystems/btrfs.rb]: So now there is a chance that `#create_btrfs_subvolume` returns a subvolume that it didn't actually create. Please, document that. It's completely unexpected to me.

File: src/lib/y2storage/filesystems/btrfs.rb (hunk 1 above)
Comment (wfeldt) [reply src/lib/y2storage/filesystems/btrfs.rb]: A quick (not representative) check shows it's not consistent; not even your comments within a class cf. `fake_device_factory.rb`.

File: src/lib/y2storage/filesystems/btrfs.rb [hunk 3]
```diff
@@ -80,6 +80,36 @@ def supports_btrfs_subvolumes?
...
+      # List of subvolumes hierarchically below path
+      #
+      # That is, subvolumes starting with path.
+      #
+      # @param path [String] subvolume path
+      #
+      # @return [Array<BtrfsSubvolume>]
+      #
+      def subvolume_descendants(path)
+        path += "/"
```
Comment (ancorgs) [inline src/lib/y2storage/filesystems/btrfs.rb]: I know we never use paths like "/var/" (trailing slash) internally. But I fear something like that can end up here via autoyast profile or lack of validation in the Partitioner. So better be safe with something like
```ruby
path += "/" unless path.end_with?("/")
```

File: src/lib/y2storage/filesystems/btrfs.rb [hunk 4]
```diff
@@ -391,6 +432,98 @@ def path_without_prefix(subvolume_path)
...
+
+      # Find the most suitable parent for a new subvolume
+      #
+      # That is, the one with the longest path that is part of path.
+      #
+      # @param path [String] subvolume path
+      #
+      # @return [BtrfsSubvolume] parent subvolume
+      #
+      def best_match(path)
```
Comment (ancorgs) [inline src/lib/y2storage/filesystems/btrfs.rb]: I'm the less suitable person to complain about names but `Btrfs#best_match` doesn't really look like the best name for a method to search a candidate parent.

File: src/lib/y2storage/filesystems/btrfs.rb (hunk 3 above)
Comment (wfeldt) [reply src/lib/y2storage/filesystems/btrfs.rb]: If that's the case the path should be put into a canonical form elsewhere. (I actually expected this.) Otherwise you'd have to deal with strange forms like `/foo///bar/` or whatever all over the place. At the very least provide a method for it.

BTW, just checked, and the path above is acceptable to the partitioner and it fails miserably when committing the change.

File: test/y2storage/filesystems/btrfs_test.rb [hunk 5]
```diff
@@ -194,25 +194,114 @@
...
+      end
+
+      it "can insert a subvolume into an existing hierarchy" do
+        filesystem.create_btrfs_subvolume(path2, nocow)
+        filesystem.create_btrfs_subvolume(path3, nocow)
+        expect(filesystem.btrfs_subvolumes.map(&:path)).to include(path3)
+      end
+    end
+
+    context "when the filesystem already exists" do
```
Comment (ancorgs) [inline test/y2storage/filesystems/btrfs_test.rb]: It also makes sense to have another test checking that it returns nil if the subvolume couldn't be inserted

File: src/lib/y2storage/filesystems/btrfs.rb [hunk 6]
```diff
@@ -80,6 +80,36 @@ def supports_btrfs_subvolumes?
...
+      #
+      # If the file system exists already there must at least be nothing
+      # else below path.
+      #
+      # @param path [String] subvolume path
+      #
+      # @return [Boolean]
+      #
+      def subvolume_can_be_created?(path)
+        return true unless exists_in_probed?
```
Comment (ancorgs) [inline src/lib/y2storage/filesystems/btrfs.rb]: I prefer to restrict the usage of `#exists_in_probed?` as much as possible. In general we would like to have code that can be used to calculate stuff taking any devicegraph as starting point, not only TheOneAndOnlyProbed . It's more testable and flexible.

For example, [BlkFilesystem#to_be_formatted?](https://github.com/yast/yast-storage-ng/blob/21a499e61698ffcf4d73d1ad532e501062371dea/src/lib/y2storage/blk_device.rb#L396) gets the initial devicegraph as argument. In the Partitioner it will get `Y2Partitioner::DeviceGraphs.system`. In the proposal that would be whatever was the starting point for the proposal (which strictly speaking doesn't HAVE to be probed). In a test it can be another thing.

Comment (wfeldt) [reply src/lib/y2storage/filesystems/btrfs.rb]: `#create_btrfs_subvolume` doesn't know a devicegraph. So `in_probed?` is the best to do, and the semantics is correct. It compares to the situation the devicegraph commit will encounter later.

File: src/lib/y2storage/filesystems/btrfs.rb (hunk 1 above)
Comment (ancorgs) [reply src/lib/y2storage/filesystems/btrfs.rb]: `fake_device_factory.rb` is one of the classes mainly created and documented by @shundhammer, who also doesn't follow the Ruby convention. As said, I'm fine with some people not following it. I just want to avoid that everyone touches what others did to adjust it to his version.

File: test/y2storage/filesystems/btrfs_test.rb [hunk 7]
```diff
@@ -230,15 +230,33 @@
...
       it "can not insert a subvolume into an existing hierarchy" do
         filesystem.create_btrfs_subvolume(path2, nocow)
-        filesystem.create_btrfs_subvolume(path3, nocow)
+        result = filesystem.create_btrfs_subvolume(path3, nocow)
+        expect(result).to be_nil
         expect(filesystem.btrfs_subvolumes.map(&:path)).to_not include(path3)
       end
+
+      it "and returns nil" do
+        expect(result).to be_nil
```
Comment (ancorgs) [inline test/y2storage/filesystems/btrfs_test.rb]: This test is doing nothing. `#create_btrfs_subvolume` has not been called here at all. I don't think you got how `let` works.

File: test/y2storage/filesystems/btrfs_test.rb [hunk 8]
```diff
@@ -230,15 +230,33 @@
...
+      end
+    end
+  end
+
+  describe "#canonical_subvolume_name" do
+    let(:devicegraph) { Y2Storage::StorageManager.instance.staging }
```
Comment (ancorgs) [inline test/y2storage/filesystems/btrfs_test.rb]: Why is this `let(:devicegraph)` here? I fail to see what is used for

File: test/y2storage/filesystems/btrfs_test.rb (hunk 5 above)
Comment (ancorgs) [reply test/y2storage/filesystems/btrfs_test.rb]: This is probably what you wanted to express:

```ruby
context "when the filesystem already exists and the subvolume does not fit in the existing hierarchy" do
  before do
    allow(filesystem).to receive(:exists_in_probed?) .and_return(true)
    filesystem.create_btrfs_subvolume(path2, nocow)
  end

  it "does not create the subvolume" do
    filesystem.create_btrfs_subvolume(path3, nocow)
    expect(filesystem.btrfs_subvolumes.map(&:path)).to_not include(path3)
  end

  it "returns nil" do
    result filesystem.create_btrfs_subvolume(path3, nocow)
    expect(result).to be_nil
  end
end
```

**Comment by ancorgs** [comment]
LGTM. This is ready for the changelog entry and version bump
